examples/drivers/windows/otLwf/tunnel.c

```diff
@@ -0,0 +1,1745 @@
+    NetBuffer->ProtocolReserved[0] = (PVOID)NetBufferLength;
+    NetBuffer->DataLength = 0;
+
+    // Pack the header, command and key
+    PackedLength = 
+        spinel_datatype_pack(
+            DataBuffer, 
+            NetBufferLength, 
+            "Cii", 
+            (spinel_tid_t)(SPINEL_HEADER_FLAG | SPINEL_HEADER_IID_0), 
+            (UINT)SPINEL_CMD_PROP_VALUE_SET, 
+            (Secured ? SPINEL_PROP_STREAM_NET : SPINEL_PROP_STREAM_NET_INSECURE));
+    if (PackedLength < 0 || PackedLength + NetBuffer->DataLength > NetBufferLength)
+    {
+        status = STATUS_INSUFFICIENT_RESOURCES;
+        goto exit;
+    }
+
+    NT_ASSERT(PackedLength >= 3);
+    NetBuffer->DataLength += (ULONG)PackedLength;
+    
+    // Copy over the data length
+    DataBuffer[NetBuffer->DataLength+1] = (((USHORT)IpNetBuffer->DataLength) >> 8) & 0xff;
+    DataBuffer[NetBuffer->DataLength]   = (((USHORT)IpNetBuffer->DataLength) >> 0) & 0xff;
+    NetBuffer->DataLength += 2;
+    
+    v6Header = (IPV6_HEADER*)(DataBuffer + NetBuffer->DataLength);
+
+    // Copy the IP packet data
+    IpDataBuffer = (PUCHAR)NdisGetDataBuffer(IpNetBuffer, IpNetBuffer->DataLength, v6Header, 1, 0);
+    if (IpDataBuffer != (PUCHAR)v6Header)
+    {
+        RtlCopyMemory(v6Header, IpDataBuffer, IpNetBuffer->DataLength);
+    }
+
+    NetBuffer->DataLength += IpNetBuffer->DataLength;
+                                            
+    LogVerbose(DRIVER_DATA_PATH, "Filter: %p, IP6_SEND: %p : %!IPV6ADDR! => %!IPV6ADDR! (%u bytes)", 
+                pFilter, NetBufferList, &v6Header->SourceAddress, &v6Header->DestinationAddress, 
+                NET_BUFFER_DATA_LENGTH(IpNetBuffer));
+
+    // Send the NBL down
+    NdisFSendNetBufferLists(
+        pFilter->FilterHandle, 
+        NetBufferList, 
+        NDIS_DEFAULT_PORT_NUMBER, 
+        DispatchLevel ? NDIS_SEND_FLAGS_DISPATCH_LEVEL : 0);
+
+    // Clear local variable because we don't own the NBL any more
+    NetBufferList = NULL;
+
+exit:
+
+    if (NetBufferList)
+    {
+        if (NetBuffer)
+        {
+            NetBuffer->DataLength = (ULONG)(ULONG_PTR)NetBuffer->ProtocolReserved[0];
+            NdisAdvanceNetBufferDataStart(NetBuffer, NetBuffer->DataLength, TRUE, NULL);
+        }
+        NdisFreeNetBufferList(NetBufferList);
+    }
+
+    return status;
+}
+
+_IRQL_requires_max_(PASSIVE_LEVEL)
+spinel_tid_t
+otLwfGetNextTunnelTransactionId(
+    _In_ PMS_FILTER pFilter
+    )
+{
+    spinel_tid_t TID = 0;
+    while (TID == 0)
+    {
+        NdisAcquireSpinLock(&pFilter->tunCommandLock);
+
+        if (((1 << pFilter->tunNextTID) & pFilter->tunTIDsInUse) == 0)
+        {
+            TID = pFilter->tunNextTID;
+            pFilter->tunNextTID = SPINEL_GET_NEXT_TID(pFilter->tunNextTID);
+        }
+
+        NdisReleaseSpinLock(&pFilter->tunCommandLock);
+
+        if (TID == 0)
+        {
+            // TODO - Wait for event
+        }
+    }
+    return TID;
```

Comment:
Actually, you are right. There is a bug here I believe. I don't see `pFilter->cmdTIDsInUse` being written to.